src/dask_awkward/lib/optimize.py

```diff
@@ -97,166 +59,44 @@ def _projectable_io_layer_names(dsk: HighLevelGraph) -> list[str]:
-
-    last_layer = list(dsk.layers.values())[-1]
-    if hasattr(last_layer, "_meta"):
-        out_meta = last_layer._meta
-        out_meta_columns = out_meta.layout.form.columns()
-        if out_meta_columns == [""]:
-            out_meta_columns = []
-    else:
-        out_meta_columns = []
-    # can only select output columns that exist in the input
-    # (other names may have come from aliases)
-
-    result: dict[str, list[str] | None] = {}
-    for pio_layer_name in pio_layer_names:
-        columns = dsk.layers[pio_layer_name]._meta.fields  # type: ignore
-        starting_columns = set(columns)
-        keep = {c for c in out_meta_columns if c in columns}
-
-        cols_used_in_getitem = set()
-        # dependencies of the current IOLayer
-        pio_layer_deps = dsk.dependents[pio_layer_name]
-        # which of those dependencies are operator.getitem layers
-        deps_that_are_getitem = [
-            k for k in pio_layer_deps if _is_getitem(dsk.layers[k])
-        ]
-        # of the getitem dependencies, determine the columns that were requested.
-        for dep_that_is_getitem in deps_that_are_getitem:
-            layer_of_interest = dsk.layers[dep_that_is_getitem]
-            requested_and_starting = (
-                _requested_columns_getitem(layer_of_interest) & starting_columns
-            )
-            requested_or_starting_and_all_calls = requested_and_starting | (
-                starting_columns & all_getitem_call_columns
-            )
-            cols_used_in_getitem |= requested_or_starting_and_all_calls
-        # project columns using the discovered getitem columns.
-        if cols_used_in_getitem:
-            keep = cols_used_in_getitem | set(keep) | set(out_meta_columns)
-            if keep == set(out_meta_columns) or keep == set(starting_columns):
-                result[pio_layer_name] = None
-            else:
-                result[pio_layer_name] = list(keep)
-
-    return result
-
-
-def _optimize_iolayer_columns_getitem(dsk: HighLevelGraph) -> HighLevelGraph:
-
-    layers_and_cols = _layers_and_columns_getitem(dsk)
-
-    # if the task graph doesn't contain a column-projectable
-    # AwkwardIOLayer then bail on this optimization (just return the
-    # existing task graph).
-    if not layers_and_cols:
-        return dsk
-
+def optimize_columns(dsk: HighLevelGraph, outputs: Iterable[str]):
+    reports = _get_column_report(dsk, outputs)
+    return _apply_column_optimization(dsk, reports)
+
+
+def _get_column_report(dsk: HighLevelGraph, outputs: Iterable[str]) -> Dict[str, Any]:
+    import awkward as ak
+    layers = dsk.layers.copy()
+    deps = dsk.dependencies.copy()
+    reports = {}
+    for name in _projectable_io_layer_names(dsk):
+        layers[name], report = layers[name].mock()
+        reports[name] = report
+    hlg = HighLevelGraph(layers, deps)
+    # TODO: outlayers should be any place we need all the columns, whether output
+    #  or to_* where data leaves dak
```

Comment:
as far as I can tell, `outputs` (passed by dask) is equal to `list(hlg.layers.values())[-1].keys()` in all the cases I can see.

Reply:
Yea I think that's the point of the dask optimization API taking two arguments: the graph and the final necessary key(s). Are you using `list(hlg.layers.values())[-1].keys()` because we had that in the previous implementation?